### PR TITLE
Add infrastructure for restricting clothing mods

### PR DESCRIPTION
**Pull request by Davi-DeGanne**
#### Summary

```SUMMARY: Infrastructure "Can set up clothing mods to be valid only for certain items"```

#### Purpose of change
Allows more granularity for clothing mods.

Required infrastructure for #34593 

#### Describe the solution
- Added the boolean field `"restricted"` to clothing mod JSONs (defaults to false)
- Added the `vector<string>` field `"valid_mods"` to armor JSONs (defaults to empty list)
- Made it so clothing mods with `"restricted": true` can only be applied to armor which lists that clothing mod's ID in `"valid_mods"`
- To make it clear to the user why mods were not applicable in this case, I had to shuffle things around a bit, and while doing so, it only made sense to explain why a mod couldn't be applied in the other instances as well; appropriate messages now display in the clothing modification menu not only when the new restrictions make a modification impossible, but also when the player doesn't have enough material or thread.
- Added documentation to JSON_INFO.md detailing the new properties.

#### Describe alternatives you've considered
Discussed in #34593, this seemed the best solution.

#### Testing
- Temporarily add `"restricted": true` to one of the clothing mods (I'll use `kevlar_padded` as an example).
- Temporarily add `"valid_mods": ["kevlar_padded"]` to any clothing item(s).
- Load the game and spawn in a tailor's kit, 100 Kevlar plates, and the clothing item(s) you modified.
- Observe that the clothing item(s) you modified can be padded with Kevlar, but all other clothing cannot.

#### Additional context
NB: Does *not* affect existing clothing modifications (as of yet), since `"restricted"` defaults to `false`.

Screenshot of UI changes:
image

### CODE REVIEW

**Comment by snipercup** [comment]
I also tested it. Works as designed:
image


I'll start making a whitelist for the upcoming steel padded mod.

File: src/iuse_actor.cpp [hunk 1]
```diff
@@ -4548,24 +4545,41 @@ int sew_advanced_actor::use( player &p, item &it, bool, const tripoint & ) const
...
-        bool enab = can_add_mod( obj.flag, obj.item_string );
+
+        bool enab = false;
         std::string prompt;
         if( mod.item_tags.count( obj.flag ) == 0 ) {
-            prompt = string_format( "%s (%d %s)", obj.implement_prompt, items_needed,
-                                    item::nname( obj.item_string, items_needed ) );
+            // Mod not already present, check if modification is possible
+            if( it.charges < thread_needed ) {
+                prompt = string_format( "Can't %s (need %d thread loaded)", obj.implement_prompt, thread_needed );
```
Comment (Qrox) [inline src/iuse_actor.cpp]:
```suggestion
                //~ %1$s: modification desc, %2$d: number of thread needed
                prompt = string_format( _( "Can't %1$s (need %2$d thread loaded)" ), obj.implement_prompt, thread_needed );
```

File: src/iuse_actor.cpp [hunk 2]
```diff
@@ -4548,24 +4545,41 @@ int sew_advanced_actor::use( player &p, item &it, bool, const tripoint & ) const
...
+                prompt[6] = std::tolower( prompt[6] );
+            } else if( !has_enough[obj.item_string] ) {
+                prompt = string_format( "Can't %s (need %d %s)", obj.implement_prompt, items_needed,
```
Comment (Qrox) [inline src/iuse_actor.cpp]:
```suggestion
                //~ %1$s: modification desc, %2$d: number of items needed, %3$s: items needed
                prompt = string_format( _( "Can't %1$s (need %2$d %3$s)" ), obj.implement_prompt, items_needed,
```

File: src/iuse_actor.cpp [hunk 3]
```diff
@@ -4548,24 +4545,41 @@ int sew_advanced_actor::use( player &p, item &it, bool, const tripoint & ) const
...
+                                        item::nname( obj.item_string, items_needed ) );
+                prompt[6] = std::tolower( prompt[6] );
+            } else if( obj.restricted &&
+                       std::find( valid_mods.begin(), valid_mods.end(), obj.flag ) == valid_mods.end() ) {
+                prompt = string_format( "Can't %s (incompatible with %s)",
```
Comment (Qrox) [inline src/iuse_actor.cpp]:
```suggestion
                //~ %1$s: modification desc, %2$s: mod name
                prompt = string_format( _( "Can't %1$s (incompatible with %2$s)" ),
```

File: src/iuse_actor.cpp [hunk 4]
```diff
@@ -4548,24 +4545,41 @@ int sew_advanced_actor::use( player &p, item &it, bool, const tripoint & ) const
...
+                                        obj.implement_prompt, mod.tname( 1, false ) );
+                prompt[6] = std::tolower( prompt[6] );
+            } else {
+                // Modification is possible
+                enab = true;
+                prompt = string_format( "%s (%d %s and %d thread)", obj.implement_prompt, items_needed,
```
Comment (Qrox) [inline src/iuse_actor.cpp]:
```suggestion
                //~ %1$s: modification desc, %2$d: number of items needed, %3$s: items needed, %4$s: number of thread needed
                prompt = string_format( _( "%1$s (%2$d %3$s and %4$d thread)" ), obj.implement_prompt, items_needed,
```

File: src/iuse_actor.cpp (hunk 1 above)
Comment (DaviBones) [reply src/iuse_actor.cpp]: Isn't `_()` redundant? It is already called on `prompt` before it is used, on line 4599.

Also, just curious, why is `%1$s` better than `%s`? For code readability/clarity, or some other reason?

Comment (anothersimulacrum) [reply src/iuse_actor.cpp]: It allows reordering the strings when translating.

Comment (Qrox) [reply src/iuse_actor.cpp]: > It is already called on prompt before it is used, on line 4599.

`_()` is used to get the translation of a string, and calling it on an already formatted string usually has no effect, so the code on line 4599 is incorrect. I guess it is intended to translate `obj.destroy_prompt` on line 4583, so you may want to fix it by moving the `_()` call on line 4599 to line 4583, around `obj.destroy_prompt`.

**Comment by Fris0uman** [comment]
Do you really need the bool? You could just use `valid_mods` to get the same results: if the item has a `valid_mods` entry then it is restricted.

File: src/iuse_actor.cpp [hunk 5]
```diff
@@ -4582,7 +4601,7 @@ int sew_advanced_actor::use( player &p, item &it, bool, const tripoint & ) const
                                          format_volume( before ), volume_units_abbr(), format_volume( after ),
                                          volume_units_abbr() ), get_volume_compare_color( before, after, true ) );
 
-        tmenu.addentry_desc( index++, enab, MENU_AUTOASSIGN, _( prompt.c_str() ), desc.str() );
+        tmenu.addentry_desc( index++, enab, MENU_AUTOASSIGN, prompt.c_str(), desc.str() );
```
Comment (Qrox) [inline src/iuse_actor.cpp]:
```suggestion
        tmenu.addentry_desc( index++, enab, MENU_AUTOASSIGN, prompt, desc.str() );
```

**Comment by DaviBones** [comment]
> Do you really need the bool? You could just use `valid_mods` to get the same results: if the item has a `valid_mods` entry then it is restricted.

Well, this was created with #34593 in mind. Doing it this way, snipercup doesn't have to add `["leather_padded", "kevlar_padded", "furred", "wooled"]` to every single armor item besides the handful of items he wants to allow steel reinforcement for; instead, only `["steel_padded"]` has to be added to said handful of items. This way seems more intuitive for JSON editors in general, and means the new item additions (and existing mods, for that matter) will have `steel_padded` disabled by default unless they specify otherwise, which is the desirable imho.

**Pull request merged**